models/gif_psc_exp.cpp

```diff
@@ -0,0 +1,469 @@
+  , S_( n.S_ )
+  , B_( n.B_, *this )
+{
+}
+
+/* ----------------------------------------------------------------
+ * Node initialization functions
+ * ---------------------------------------------------------------- */
+
+void
+nest::gif_psc_exp::init_state_( const Node& proto )
+{
+  const gif_psc_exp& pr = downcast< gif_psc_exp >( proto );
+  S_ = pr.S_;
+  // S_.r_ref_ = Time(Time::ms(P_.t_ref_remaining_)).get_steps();
+}
+
+void
+nest::gif_psc_exp::init_buffers_()
+{
+  B_.spikes_ex_.clear(); // includes resize
+  B_.spikes_in_.clear(); // includes resize
+  B_.currents_.clear();  //!< includes resize
+  B_.logger_.reset();    //!< includes resize
+  Archiving_Node::clear_history();
+}
+
+void
+nest::gif_psc_exp::calibrate()
+{
+
+  B_.logger_.init();
+
+  V_.h_ = Time::get_resolution().get_ms();
+  V_.rng_ = kernel().rng_manager.get_rng( get_thread() );
+
+  V_.P11ex_ = std::exp( -V_.h_ / P_.tau_ex_ );
+  V_.P11in_ = std::exp( -V_.h_ / P_.tau_in_ );
+
+  double_t tau_m = P_.c_m_ / P_.g_L_;
+
+  // these are determined according to a numeric stability criterion
+  V_.P21ex_ = propagator_32( P_.tau_ex_, tau_m, P_.c_m_, V_.h_ );
+  V_.P21in_ = propagator_32( P_.tau_in_, tau_m, P_.c_m_, V_.h_ );
+
+
+  V_.P33_ = std::exp( -V_.h_ / tau_m );
+  V_.P30_ = 1 / P_.c_m_ * ( 1 - V_.P33_ ) * tau_m;
+  V_.P31_ = ( 1 - V_.P33_ );
+
+  V_.RefractoryCounts_ = Time( Time::ms( P_.t_ref_ ) ).get_steps();
+  assert( V_.RefractoryCounts_ >= 0 ); // since t_ref_ >= 0, this can only fail in error
+
+  // initializing internal state
+  if ( !S_.initialized_ )
+  {
+    for ( uint_t i = 0; i < P_.tau_sfa_.size(); i++ )
+    {
+      V_.Q33_.push_back( std::exp( -V_.h_ / P_.tau_sfa_[ i ] ) );
+      S_.q_sfa_elems_.push_back( 0.0 );
+    }
+
+    for ( uint_t i = 0; i < P_.tau_stc_.size(); i++ )
+    {
+      V_.Q44_.push_back( std::exp( -V_.h_ / P_.tau_stc_[ i ] ) );
+      S_.q_stc_elems_.push_back( 0.0 );
+    }
+
+    S_.initialized_ = true;
+  }
+}
+
+/* ----------------------------------------------------------------
+ * Update and spike handling functions
+ */
+
+void
+nest::gif_psc_exp::update( Time const& origin, const long_t from, const long_t to )
+{
+
+  assert( to >= 0 && ( delay ) from < kernel().connection_builder_manager.get_min_delay() );
+  assert( from < to );
+
+  double_t q_temp_;
+
+  for ( long_t lag = from; lag < to; ++lag )
+  {
+
+    q_temp_ = 0;
+    for ( uint_t i = 0; i < S_.q_stc_elems_.size(); i++ )
+    {
+
+      q_temp_ += S_.q_stc_elems_[ i ];
+
+      S_.q_stc_elems_[ i ] = V_.Q44_[ i ] * S_.q_stc_elems_[ i ];
+    }
+
+    S_.stc_ = q_temp_;
+
+    q_temp_ = 0;
+    for ( uint_t i = 0; i < S_.q_sfa_elems_.size(); i++ )
+    {
+
+      q_temp_ += S_.q_sfa_elems_[ i ];
+
+      S_.q_sfa_elems_[ i ] = V_.Q33_[ i ] * S_.q_sfa_elems_[ i ];
+    }
+
+    S_.q_ = q_temp_ + P_.v_t_star_;
+
+    ulong_t n_spikes = 0;
+
+    if ( S_.r_ref_ == 0 ) // neuron not refractory, so evolve V
+    {
+
+      if ( S_.add_stc_sfa_ == true )
```

Comment:
At this point, there is an interesting question about the precise definition of the model. In all other NEST models, refractoriness affects *membrane potential only*. In your implementation, though, you let `stc` (eta) and `sfa` (gamma) evolve during the refractory period as if the spike did not yet happen (lines 323-434), and then apply the jump to both values at the end of the refractory time. Eqs 1 and 4 in the PLoS paper gather give the impression that eta and gamma set in at the spike time t_j, not at t_j+t_ref.

Therefore, I would suggest to add the jumps right after a spike has been applied, in the code block that also emits the spike; you can then drop `S_.add_stc_sfa_`. One then needs to be a bit careful about the next time step:

1. Let a time step be `(t, t+h]` (time steps in NEST are always left open, right closed intervals). The `stc_elems_` at the end of a time step contain the value at the end of the step, including a jump if the neuron spiked during the step.
1. We need to use those `stc_elems_` values when computing the total `stc` entering into the membrane potential update.
1. Only after we have used the values must we update them.

You do this already correctly on lines 323-343 (the same applies to sfa)

The only problem is the sfa value returned by `State_::get()`, since that value is only updated at the beginning of the next update loop and would just not contain any jumps occurring during the last step of a `Simulate` call. The best way to make sure that the correct sfa value is returned, is probably to compute the sfa value explicity in the `get()` method.